python/sdist/amici/de_export.py

```diff
@@ -1638,6 +1646,132 @@ def free_symbols(self) -> set[sp.Basic]:
             )
         )
 
+    def static_indices(self, name: str) -> list[int]:
+        """
+        Returns the indices of static expressions in the given model entity.
+
+        Static expressions are those that do not depend on time,
+        neither directly nor indirectly.
+
+        :param name: Name of the model entity.
+        :return: List of indices of static expressions.
+        """
+        # already computed?
+        if (res := self._static_indices.get(name)) is not None:
+            return res
+
+        if name == "w":
+            dwdx = self.eq("dwdx")
+            dwdw = self.eq("dwdw")
+            w = self.eq("w")
+
+            # check for indirect state and time dependency
+            # to avoid lengthy symbolic computations,
+            # we only check if we have any non-zeros in hierarchy,
+            # we currently neglect the case where different hierarchy
+            # levels may cancel out. treating a static expression as
+            # dynamic in such rare cases shouldn't be a problem.
+            nonzero_dwdx = np.asarray(
+                dwdx.applyfunc(lambda x: int(not x.is_zero))
+            ).astype(np.int64)
+            # to check for other time-dependence, we add a column to the dwdx
+            #  matrix
+            dynamic_syms = [
+                # FIXME: see spline comment below
+                # *self.sym("spl"),
+                *self.sym("h"),
+                amici_time_symbol,
+            ]
+            nonzero_dwdx = np.hstack(
+                (
+                    nonzero_dwdx,
+                    np.array(
+                        [
+                            expr.has(*dynamic_syms)
+                            # FIXME: the current spline implementation is a giant pita
+                            #  currently, the splines occur in the form of sympy functions, e.g.:
+                            #   AmiciSpline(y0, time, y0_3, y0_1)
+                            #   AmiciSplineSensitivity(y0, time, y0_1, y0_3, y0_1)
+                            #  until it uses the proper self.sym("spl") / self.sym("sspl")
+                            #  symbols, which will require quite some refactoring,
+                            #  we just do dumb string checks :|
+                            or (
+                                bool(self._splines)
+                                and "AmiciSpline" in str(expr)
+                            )
+                            for expr in w
+                        ]
+                    )[:, np.newaxis],
+                )
+            )
+
+            nonzero_dwdw = np.asarray(
+                dwdw.applyfunc(lambda x: int(not x.is_zero))
+            ).astype(np.int64)
+
+            tmp = nonzero_dwdx
```

Comment:
add comment that this unravels the hierarchical structure of `w` and that `np.sum(tmp, axis=1)>0` checks for direct/indirect time dependency?